Universal FE Randomizer/src/random/gba/randomizer/service/GBASlotAdjustmentService.java

```diff
@@ -0,0 +1,183 @@
+package random.gba.randomizer.service;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Random;
+import java.util.stream.Collectors;
+
+import fedata.gba.GBAFECharacterData;
+import fedata.gba.GBAFEClassData;
+import fedata.gba.GBAFEStatDto;
+import random.gba.loader.ClassDataLoader;
+import random.gba.loader.TextLoader;
+import ui.model.RecruitmentOptions;
+import ui.model.RecruitmentOptions.ClassMode;
+import util.DebugPrinter;
+
+public class GBASlotAdjustmentService {
+
+	private static final int ASSUMED_PROMOTION_LEVEL = 10;
+	
+	/**
+	 * Used by Recruitment Randomization and Character Shuffling to Calculate the following information:
+	 * 
+	 * A) Should the character, be demoted, promoted or are they already in the right tier of class.
+	 * B) How many auto levels should they receive
+	 * C) Which promotion bonuses will they receive
+	 */
+	public static ClassAdjustmentDto handleClassAdjustment(int targetLevel, int sourceLevel,
+			boolean shouldBePromoted, boolean isPromoted, Random rng, ClassDataLoader classData, 
+			GBAFEClassData targetClass, GBAFEClassData fillSourceClass, GBAFECharacterData fill, 
+			GBAFEClassData slotSourceClass, RecruitmentOptions options, TextLoader textData, DebugPrinter.Key key) {
+		ClassAdjustmentDto dto = new ClassAdjustmentDto();
+		if (shouldBePromoted) { targetLevel += ASSUMED_PROMOTION_LEVEL; }
+		if (isPromoted) { sourceLevel += ASSUMED_PROMOTION_LEVEL; }
+		dto.levelAdjustment = targetLevel - sourceLevel;
+		
+		// To make newly created pre-promotes not completely busted (since they probably had higher growths than real pre-promotes)
+		// we'll subtract a few levels from their autoleveling amount, assuming they get a lot (like 10).
+		if (!isPromoted && shouldBePromoted && dto.levelAdjustment > 10) {
+			DebugPrinter.log(key, "Dropping 3 additional levels for new prepromotes.");
+			dto.levelAdjustment  -= 3;
+		} else if(isPromoted && !shouldBePromoted && dto.levelAdjustment < -10) {
```

Comment:
Should these use the `ASSUMED_PROMOTION_LEVEL`?

Reply:
I don't think they should as it doesn't represent the same information, and I certainly could see this 10 changing independently of the ASSUMED_PROMOTION_LEVEL.

This check was added since f.e. a Knoll randomizing into Seth would by the old logic get -2 auto levels (newly prepromoted so -3) And I wanted to apply that only if they got a significant number of levels. 10 was just an initial suggestion, would have to see what works best for this.

Reply:
Yeah, the old auto leveling routine always needed some tweaking, but I could never find a balanced way of doing it. As it usually made some characters completely busted or completely useless. If these are levers that you want to change later, it's probably worth coming up with a new constant at least to represent what they are. It makes it a bit easier to reason about.

Reply:
Definetly makes sense you are right, I would change it to this then:
![2023-03-26 09_43_35-Yune_Trunk - Universal FE Randomizer_src_random_gba_randomizer_service_GBASlotAd](https://user-images.githubusercontent.com/41942612/227762289-e4091bed-4b17-4d66-b9a0-2bebb3f6883d.png)